tools/protoc-gen-docs/htmlGenerator.go

```diff
@@ -0,0 +1,916 @@
+		}
+		g.emit("</code></pre>")
+
+		g.generateComment(method.loc, method.GetName())
+	}
+
+	g.generateSectionTrailing()
+}
+
+// emit prints the arguments to the generated output.
+func (g *htmlGenerator) emit(str ...string) {
+	for _, s := range str {
+		g.buffer.WriteString(s)
+	}
+	g.buffer.WriteByte('\n')
+}
+
+var typeLinkPattern = regexp.MustCompile("\\[.*\\]\\[.*\\]")
+
+func (g *htmlGenerator) generateComment(loc *descriptor.SourceCodeInfo_Location, name string) {
+	com := loc.GetLeadingComments()
+	if com == "" {
+		com = loc.GetTrailingComments()
+		if com == "" {
+			g.warn(loc, "no comment found for %s", name)
+			return
+		}
+	}
+
+	text := strings.TrimSuffix(com, "\n")
+	lines := strings.Split(text, "\n")
+	if len(lines) > 0 {
+		// Based on the amount of spacing at the start of the first line,
+		// remove that many bytes at the beginning of every line in the comment.
+		// This is so we don't inject extra spaces in any preformatted blocks included
+		// in the comments
+		pad := 0
+		for i, ch := range lines[0] {
+			if !unicode.IsSpace(ch) {
+				pad = i
+				break
+			}
+		}
+
+		for i := 0; i < len(lines); i++ {
+			l := lines[i]
+			if len(l) > pad {
+				lines[i] = l[pad:]
+			}
+		}
+
+		// now, adjust any headers included in the comment to correspond to the right
+		// level, based on the heading level of the surrounding content
+		for i := 0; i < len(lines); i++ {
+			l := lines[i]
+			if strings.HasPrefix(l, "#") {
+				if g.grouping {
+					lines[i] = "##" + l
+				} else {
+					lines[i] = "#" + l
+				}
+			}
+		}
+
+		// find any type links of the form [name][type] and turn
+		// them into normal HTML links
+		for i := 0; i < len(lines); i++ {
+
+			lines[i] = typeLinkPattern.ReplaceAllStringFunc(lines[i], func(match string) string {
+				end := 0
+				for match[end] != ']' {
+					end++
+				}
+
+				linkName := match[1:end]
+				typeName := match[end+2 : len(match)-1]
+
+				if o, ok := g.model.allDescByName["."+typeName]; ok {
+					return g.linkify(o, linkName)
+				}
+
+				if l, ok := wellKnownTypes[typeName]; ok {
+					return "<a href='" + l + "'>" + linkName + "</a>"
+				}
+
+				return "*" + linkName + "*"
+			})
+		}
+	}
+
+	text = strings.Join(lines, "\n")
+
+	// turn the comment from markdown into HTML
+	result := blackfriday.Run([]byte(text))
+
+	g.buffer.Write(result)
+	g.buffer.WriteByte('\n')
+}
+
+// well-known types whose documentation we can refer to
+var wellKnownTypes = map[string]string{
+	"google.protobuf.Duration":               "https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Duration",
+	"google.protobuf.Timestamp":              "https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Timestamp",
+	"google.api.MetricDescriptor.MetricKind": "https://cloud.google.com/logging/docs/reference/v2/rest/v2/projects.metrics#metrickind",
```

Comment:
I am not sure if this is a good idea to reference logging api reference docs.. also they are rest format showcasing json style example, whereas most of the other messages that we are emitting are proto style of documentation.